File: src/s3vs_lib/MouseInterpreter.cpp
/*
ReVisE: Remote visualization environment for large datasets
Copyright (C) 2021 Stepan Orlov, Alexey Kuzin, Alexey Zhuravlev, Vyacheslav Reshetnikov, Egor Usik, Vladislav Kiev, Andrey Pyatlin

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see https://www.gnu.org/licenses/agpl-3.0.en.html.

*/

#include "MouseInterpreter.hpp"
#include "CameraController.hpp"

using namespace s3dmm;

namespace s3vs
{

MouseInterpreter::MouseInterpreter(
        const SyncVsControllerInput& input,
        CameraController &cameraController,
        const CameraTransformSetter& cameraTransformSetter) :
    m_input(input),
    m_cameraController(cameraController),
    m_cameraTransformSetter(cameraTransformSetter)
{
}

void MouseInterpreter::interpretMouse(const MouseState& mouseState)
{
    trackClickTime(mouseState);
    switch (mouseState.flags) {
    case MouseState::LeftButton:
        if (mouseMoved(mouseState) && (m_prevMouseState.flags & MouseState::LeftButton))
            rotate(mouseState);
        break;
    case MouseState::MiddleButton:
        if (mouseMoved(mouseState) && (m_prevMouseState.flags & MouseState::MiddleButton))
            pan(mouseState);
        break;
    case MouseState::RightButton:
    case MouseState::RightButton + MouseState::ShiftKey:
        if (mouseMoved(mouseState) && (m_prevMouseState.flags & MouseState::RightButton))
            zoomByMouseMove(mouseState);
        break;
    case 0:
        if (mouseDoubleClicked()) {
            resetCameraTransform();
            break;
        }
        // Intentionally falling through
    case MouseState::ShiftKey:
        if (mouseState.wheelDelta != 0)
            zoomByMouseWheel(mouseState);
        break;
    }
    m_prevMouseState = mouseState;
}

void MouseInterpreter::trackClickTime(const MouseState& mouseState)
{
    m_mouseClicked = m_mouseDoubleClicked = false;
    auto now = std::chrono::steady_clock::now();
    if (m_prevMouseState.flags == 0 && mouseState.flags == MouseState::LeftButton) {
        // Left mouse button has been pressed
        m_prevPrevLButtonDown = m_prevLButtonDown;
        m_prevLButtonDown = now;
    }
    else if (mouseState.flags != 0 && mouseState.flags != MouseState::LeftButton) {
        // Something has happened that destroys the click
        auto longAgo = now - std::chrono::seconds(10);
        m_prevLButtonDown = m_prevPrevLButtonDown = longAgo;
    }
    else if (m_prevMouseState.flags == MouseState::LeftButton && mouseState.flags == 0) {
        // Left mouse button has been released
        constexpr std::chrono::milliseconds maxClickDuration(300);
        constexpr std::chrono::milliseconds maxDoubleClickDuration(500);
        m_mouseClicked =
            std::chrono::duration_cast<std::chrono::milliseconds>(now - m_prevLButtonDown)
            < maxClickDuration;
        m_mouseDoubleClicked =
            std::chrono::duration_cast<std::chrono::milliseconds>(now - m_prevPrevLButtonDown)
            < maxDoubleClickDuration;
    }
}

void MouseInterpreter::rotate(const MouseState& mouseState)
{
    m_cameraController.rotate(mouseDr(mouseState));
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

void MouseInterpreter::rotateInPlane(const MouseState& mouseState)
{
    m_cameraController.rotateInPlane(mouseDr(mouseState));
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

void MouseInterpreter::pan(const MouseState& mouseState)
{
    m_cameraController.pan(mouseDr(mouseState));
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

void MouseInterpreter::zoomByMouseWheel(const MouseState &mouseState)
{
    constexpr auto ShiftZoomFactor = make_real(0.1);
    auto delta = make_real(mouseState.wheelDelta)/120;
    if (mouseState.flags & MouseState::ShiftKey)
        delta *= make_real(ShiftZoomFactor);
    m_cameraController.zoom(mousePos(mouseState), delta);
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

void MouseInterpreter::zoomByMouseMove(const MouseState& mouseState)
{
    constexpr auto ShiftZoomFactor = make_real(0.1);
    auto delta = mouseDr(mouseState)[1] / make_real(100);
    if (mouseState.flags & MouseState::ShiftKey)
        delta *= make_real(ShiftZoomFactor);
    m_cameraController.zoom(mousePos(mouseState), delta);
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

void MouseInterpreter::resetCameraTransform() {
    m_cameraController.resetCameraTransform();
    m_cameraTransformSetter(m_cameraController.cameraTransform());
}

Vec2r MouseInterpreter::mousePos(const MouseState& mouseState) const
{
    return {
        make_real(mouseState.x),
        make_real(m_input.access()->viewportSize()[1] - mouseState.y)
    };
}

Vec2r MouseInterpreter::mouseDr(const MouseState& mouseState) const
{
    return {
        make_real(mouseState.x - m_prevMouseState.x),
       -make_real(mouseState.y - m_prevMouseState.y)
    };
}

bool MouseInterpreter::mouseMoved(const MouseState& mouseState) const
{
    return !(mouseState.x == m_prevMouseState.x && mouseState.y == m_prevMouseState.y);
}

bool MouseInterpreter::mouseClicked() const {
    return m_mouseClicked;
}

bool MouseInterpreter::mouseDoubleClicked() const {
    return m_mouseDoubleClicked;
}

} // namespace s3vs

File: src/webserver/public/javascripts/components/ctf/MouseInterpreter.js
/*
ReVisE: Remote visualization environment for large datasets
Copyright (C) 2021 Stepan Orlov, Alexey Kuzin, Alexey Zhuravlev, Vyacheslav Reshetnikov, Egor Usik, Vladislav Kiev, Andrey Pyatlin

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see https://www.gnu.org/licenses/agpl-3.0.en.html.

*/

(function(){
    let ctf = objects.Ctf;

    class MouseInterpreter {
        constructor() {
            this.mouseClicked = false;
            this.mousein = false;
            
            this.canvas = null;
            this.scene = null;
            this.helper = null;

            this.canvasPosition = { top: 0,
                                    left: 0 };
        }
        setCanvas(canvas) {
            let self = this;
            let jquery = canvas.jquery;
            jquery.mouseover(e => self.mousein = true);
            jquery.mouseleave(e => self.mousein = false);
            jquery.mousedown(e => {
                self.canvasPosition = {
                    top: e.pageY - e.offsetY,
                    left: e.pageX - e.offsetX
                };

                self.mouseClicked = true;
                self.onMouseClicked(e.offsetX, e.offsetY)
            });
            $(document).mouseup(e => {
                if(self.mouseClicked) {
                    let pos = self.canvasPosition;
                    let x = e.pageX - pos.left;
                    let y = e.pageY - pos.top;
                    self.mouseClicked = false;
                    self.onMouseReleased(x, y);
		}
            });
            $(document).mousemove(e => {
	       	if(self.mouseClicked) {
                    let pos = self.canvasPosition;
                    let x = e.pageX - pos.left;
                    let y = e.pageY - pos.top;
                    let dx = e.originalEvent.movementX;
                    let dy = e.originalEvent.movementY;
                    self.onMouseMoved(x, y, dx, dy);
		}
            });

        }
        setScene(scene) {
            this.scene = scene;
        }
        attachHelper(helper) {
            this.helper = helper;
        }
    }
    ctf.MouseInterpreter = MouseInterpreter;


    class OpacityMouseInterpreter extends MouseInterpreter {
        constructor() {
            super();
            this.draggedPoint = null;
            this.selectedPoint = null;
        }
        onMouseClicked(x, y) {
            if(this.selectedPoint)
                this.selectedPoint.selected = false;
            this.selectedPoint = null;
            let p = this.scene.findByScreenCoordinates(x, y);
            if(!p) {
                let sceneX = this.helper.coordinatesTransform.fromScreenX(x);
                let sceneY = this.helper.coordinatesTransform.fromScreenY(y);
                p = this.scene.createPoint(sceneX, sceneY);
            } else {
                this.scene.setSelected(p);
                this.scene.repaint();
            }
            this.selectedPoint = p;
            this.draggedPoint = p;
        }
        onMouseMoved(x, y, dx, dy) {
            if(this.draggedPoint) {
	        this.scene.moveSelected(
                    this.helper.fromScreenX(x), 
                    this.helper.fromScreenY(y));
                this.scene.repaint();
	    }
           
        }
        onMouseReleased(x, y) {
            if(this.draggedPoint) {
                this.draggedPoint = null;
                this.scene.dispatchEvent(new Event("change"));
            }
        }
    }
    ctf.OpacityMouseInterpreter = OpacityMouseInterpreter;


    class ColormapMouseInterpreter extends MouseInterpreter {
        constructor() {
            super();
            this.selectedPoint = null;
            this.draggedPoint = null;
        }
        onMouseClicked(x, y) {
            if(!this.mouseClicked)
                return;
            if(this.selectedPoint)
                this.selectedPoint.selected = false;
            this.selectedPoint = null;
            let p = this.scene.findByScreenCoordinates(x, y);
            if(!p) {
                let sceneX = this.helper.coordinatesTransform.fromScreenX(x);
                let sceneY = this.helper.coordinatesTransform.fromScreenY(y);

                p = this.scene.createPoint(sceneX, sceneY);
            } else {
                this.scene.setSelected(p);
                this.scene.repaint();
            }
            this.selectedPoint = p;
            this.draggedPoint = p;
        }
        onMouseMoved(x, y, dx, dy) {
            if(this.draggedPoint) {
                let x_ = this.helper.fromScreenX(x);
                let y_ = this.helper.fromScreenY(y);
                this.scene.moveSelected(x_, y_);
                this.scene.repaint();
	    }
        }
        onMouseReleased(x, y) { 
            if(this.draggedPoint) {
                this.draggedPoint = null;
                this.scene.dispatchEvent(new Event("change"));
            }
        }
    }
    ctf.ColormapMouseInterpreter = ColormapMouseInterpreter;
})()
